src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                        {
+                            if (byValue.ValueAndDecision.Count != 0) return null; // can't tell which decision to take
+                        }
+                        else
+                        {
+                            foreach (var vd in byValue.ValueAndDecision)
+                            {
+                                var value = vd.Key;
+                                var decision = vd.Value;
+                                if (Equals(byValue.Expression.ConstantValue.Value, value)) return BindConstantTarget(decision);
+                            }
+                        }
+                        return (byValue.Default != null) ? BindConstantTarget(byValue.Default) : null;
+                    }
+                case DecisionTree.DecisionKind.Guarded:
+                    {
+                        var guarded = (DecisionTree.Guarded)decisionTree;
+                        return (guarded.Guard == null || guarded.Guard.ConstantValue == ConstantValue.True) ? guarded.Label.Label : null;
+                    }
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(decisionTree.Kind);
+            }
+        }
+
+        private void EnsureDecisionTree()
+        {
+            if (_decisionTree == null)
+            {
+                // it is not a problem if this is done concurrently by multiple threads, as they
+                // should all compute semantically equivalent diagnostics.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var decisionTree = ComputeDecisionTree(diagnostics);
+                this._decisionTreeDiagnostics = diagnostics.ToReadOnlyAndFree();
+                Interlocked.CompareExchange(ref _decisionTree, decisionTree, null);
+            }
+        }
+
+        private DecisionTree ComputeDecisionTree(DiagnosticBag diagnostics)
+        {
+            DecisionTreeComputer decisionTreeComputer = new DecisionTreeComputer(
+                this.Binder.ContainingMemberOrLambda, diagnostics, this, this.Binder.Conversions);
+            return decisionTreeComputer.ComputeDecisionTree();
+        }
+    }
+
+    internal class DecisionTreeComputer
+    {
+        private readonly Symbol _enclosingSymbol;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly BoundPatternSwitchStatement _switchStatement;
+        private readonly Conversions _conversions;
+
+        private CSharpSyntaxNode _syntax;
+
+        public DecisionTreeComputer(
+            Symbol enclosingSymbol,
+            DiagnosticBag diagnostics,
+            BoundPatternSwitchStatement switchStatement,
+            Conversions conversions)
+        {
+            this._enclosingSymbol = enclosingSymbol;
+            this._diagnostics = diagnostics;
+            this._switchStatement = switchStatement;
+            this._conversions = conversions;
+        }
+
+        internal DecisionTree ComputeDecisionTree()
```

Comment:
I will look at simplifying this code as it evolves. Though you didn't see the previous state, the separation of subsumption checking from addition is a recent change, and I expect more changes in that direction.

---
In reply to: [64984247](https://github.com/dotnet/roslyn/pull/11485#discussion_r64984247) [](ancestors = 64984247)